Use system malloc when lite MallocStackLogging is enabled

bnham opened the pull request:
#### b6454a5cd58f928df130dfecc1627dc60f3023ab
<pre>
Use system malloc when lite MallocStackLogging is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=256675">https://bugs.webkit.org/show_bug.cgi?id=256675</a>
rdar://109233872

Reviewed by Yusuke Suzuki.

In 255601@main we added support for MallocStackLogging in libpas. This seems to work when full logging is enabled, but not when lite logging is enabled (rdar://109283870).

To work around this, force using system malloc when lite logging is enabled.

* Source/bmalloc/bmalloc/Environment.cpp:
(bmalloc::isWebKitMallocForceEnabled):
(bmalloc::isMallocEnvironmentVariableImplyingSystemMallocSet):
(bmalloc::Environment::computeIsDebugHeapEnabled):

Canonical link: <a href="https://commits.webkit.org/264075@main">https://commits.webkit.org/264075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a7f62c65e5e3e145e240e54b6aa1cdc20da2766

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6482 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8063 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6773 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6643 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9670 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5877 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8143 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4121 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5855 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5459 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6052 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5934 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8265 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6074 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6418 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5249 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6617 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5823 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1542 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1565 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9983 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6800 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6195 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1671 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/bmalloc/bmalloc/Environment.cpp:
```diff
@@ -83,6 +90,12 @@ static bool isMallocEnvironmentVariableImplyingSystemMallocSet()
+    // Use system malloc anytime MallocStackLogging is enabled, except when the "vm" or "vmlite" logging modes are enabled.
+    // Those modes only intercept syscalls rather than mallocs, so they don't necessarily imply the use of system malloc.
+    const char* mallocStackLogging = getenv("MallocStackLogging");
+    if (mallocStackLogging && strncmp(mallocStackLogging, "vm", 2))
```
I think this is not working only when `lite` is specified. So we should allow the other options.

bnham replied:
I think the style checker insisting that we not compare to 0 makes things confusing here.

`strncmp(msl, "vm", 2)` is equivalent to `strncmp(msl, "vm", 2) != 0` is equivalent to `!hasPrefix(msl, "vm")`.

So basically this is saying that if the `MallocStackLogging` environment variable exists, and its value does not start with `vm`, then use system malloc. I think this is what we want. For instance, if `MallocStackLogging=lite`, then we will use system malloc (which is the point of this patch).

FWIW this isn't new code. I basically just reverted the Environment.cpp changes from 255601@main.

Constellation replied:
But the problem is only for `lite`, right? I can see correct memgraph if `MallocStackLogging=full`.

bnham replied:
Ok, I am not sure if memgraphs in WebContent are correct when full MSL is enabled since testing that that requires changing the WebContent sandbox. But restricting this to just lite seems fine.

Constellation reviewed (approved):
r=me

bnham committed:
Use system malloc when lite MallocStackLogging is enabled
https://bugs.webkit.org/show_bug.cgi?id=256675
rdar://109233872

Reviewed by Yusuke Suzuki.

In 255601@main we added support for MallocStackLogging in libpas. This seems to work when full logging is enabled, but not when lite logging is enabled (rdar://109283870).

To work around this, force using system malloc when lite logging is enabled.

* Source/bmalloc/bmalloc/Environment.cpp:
(bmalloc::isWebKitMallocForceEnabled):
(bmalloc::isMallocEnvironmentVariableImplyingSystemMallocSet):
(bmalloc::Environment::computeIsDebugHeapEnabled):

Canonical link: https://commits.webkit.org/264075@main
